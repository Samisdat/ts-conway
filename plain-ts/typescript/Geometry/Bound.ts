export class Bound {

    private min: number;

    private max: number;

    constructor(min: number, max: number) {

        if (max < min) {
            throw new Error('max can not be lower then min');
        }

        this.min = min;
        this.max = max;
    }

    public isWithin(test: number): boolean {
        return (test >= this.min && test <= this.max);
    }

    public isBelow(test: number): boolean {
        return (test < this.min);
    }

    public isAbove(test: number): boolean {
        return (test > this.max);
    }

    public confine(value: number): number {

        if (value < this.min) {
            return this.min;
        }

        if (value > this.max) {
            return this.max;
        }

        return value;

    }

    public getMax(): number{
        return this.max;
    }

    public getMin(): number{
        return this.min;
    }

    public expand(value: number): void{

        if(true === this.isWithin(value)){
            return;
        }

        if(true === this.isBelow(value)){
            this.min = value
        }

        if(true === this.isAbove(value)){
            this.max = value
        }

    }

}
