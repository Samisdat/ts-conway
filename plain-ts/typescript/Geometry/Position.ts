export class Position {

    public readonly  x: number;
    public readonly  y: number;

    constructor(x: number, y: number) {

        if (-0 === x) {
            x = 0;
        }

        if (-0 === y) {
            y = 0;
        }

        this.x = x;
        this.y = y;
    }

    public move(position: Position): Position {
        return new Position(
            this.x + position.x,
            this.y + position.y
        );
    }

    public inverse(): Position {
        return new Position(
            -1 * this.x,
            -1 * this.y
        );
    }

    public compare(position: Position): boolean {

        return (position.x === this.x && position.y === this.y);

    }

    public getNeighbours(): Position[] {

        const neighbours: Position[] = [];

        // neighbours on bottom
        neighbours.push(
            new Position(this.x - 1, this.y - 1)
        );
        neighbours.push(
            new Position(this.x, this.y - 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y - 1)
        );

        // neighbours on the left and on the right
        neighbours.push(
            new Position(this.x - 1, this.y)
        );
        neighbours.push(
            new Position(this.x + 1, this.y)
        );

        // neighbours on top
        neighbours.push(
            new Position(this.x - 1, this.y + 1)
        );
        neighbours.push(
            new Position(this.x, this.y + 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y + 1)
        );

        return neighbours;
    }

    public clone(): Position {
        return new Position(
            this.x,
            this.y
        );
    }

    public toString(): string {

        let str = '';

        str += (0 <= this.x) ? '+' : '';
        str += this.x;

        str += '/';

        str += (0 <= this.y) ? '+' : '';
        str += this.y;

        return str;

    }
}
