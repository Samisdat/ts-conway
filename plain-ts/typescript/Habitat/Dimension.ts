export class Dimension {

    private readonly _rows: number;
    private readonly _cols: number;

    constructor(rows: number, cols: number) {

        this.ensurePositiveNumber(rows, 'rows');
        this.ensurePositiveNumber(cols, 'cols');

        this._rows = rows;
        this._cols = cols;

    }

    public getRows(): number {
        return this._rows;
    }

    public getCols(): number {
        return this._cols;
    }

    private ensurePositiveNumber(value: number, attributeName: string) {

        if (1 > value) {
            throw new Error(attributeName + ' must be at least 1');
        }


    }
}
