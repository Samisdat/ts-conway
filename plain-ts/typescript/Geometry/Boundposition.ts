import {Position} from '@Conway/Geometry/Position';
import {Bound} from '@Conway/Geometry/Bound';

export class Boundposition {

    private x: Bound;
    private y: Bound;

    constructor(topLeft: Position, bottomRight: Position) {

        if (bottomRight.x < topLeft.x) {
            throw new Error('bottomRight must be right from topLeft');
        }

        if (bottomRight.y < topLeft.y) {
            throw new Error('bottomRight must be below from topLeft');
        }

        this.x = new Bound(topLeft.x, bottomRight.x);
        this.y = new Bound(topLeft.y, bottomRight.y);

    }

    public isWithin(test: Position): boolean {
        return (this.x.isWithin(test.x) && this.y.isWithin(test.y));
    }

    public confine(value: Position): Position {

        return new Position(
            this.x.confine(value.x),
            this.y.confine(value.y)
        );

    }

    public bottomLeft(): Position{

        return new Position(this.x.getMin(), this.y.getMin());

    }

    public topRight(): Position{

        return new Position(this.x.getMax(), this.y.getMax());

    }

    public getWidth(): number{

        return this.x.getMax() - this.x.getMin() + 1;

    }

    public getHeight(): number{

        return this.y.getMax() - this.y.getMin() + 1;

    }

    public expand(position: Position): void{

        if(true === this.isWithin(position)){
            return;
        }

        this.x.expand(position.x);
        this.y.expand(position.y);

    }

}
