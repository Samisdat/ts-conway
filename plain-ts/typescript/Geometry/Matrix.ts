import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

interface MatrixInterface {
    [index: string]: Position;
}


export class Matrix {

    private matrix: MatrixInterface = {};

    private boundposition: Boundposition;

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;


        if(undefined === this.boundposition){
            this.boundposition = new Boundposition(position, position);
        }
        else{
            this.boundposition.expand(position);
        }

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public getAll(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    public getBound(): Boundposition {

        return this.boundposition;

    }

    public getWidth(): number {

        return this.getBound().getWidth();

    }

    public getHeight(): number {

        return this.getBound().getHeight();

    }

}
