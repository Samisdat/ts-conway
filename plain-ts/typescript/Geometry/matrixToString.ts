import {Matrix} from '@Conway/Geometry/Matrix';
import {Position} from '@Conway/Geometry/Position';

interface NumberedScale {
    sign: string[];
    number: string[];
};

const getSigns = (numberedScale: NumberedScale, i: number) =>{
    if(0 === i % 5){

        numberedScale.number.push(Math.abs(i).toString());

        if(0 > i){
            numberedScale.sign.push('-');
        }
        else if(0 < i){
            numberedScale.sign.push('+');
        }
        else{
            numberedScale.sign.push(' ');
        }

    }
    else{
        numberedScale.number.push(' ');
        numberedScale.sign.push(' ');
    }

};

const getScaleHorizontal = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const left = matrix.getBound().bottomLeft().x;
    const right  = matrix.getBound().topRight().x;

    matrixArray = matrixArray.reverse();

    const newRow: string[] = []
    for(let i = 0; i < matrixArray[0].length; i += 1){
        newRow.push('')
    }

    matrixArray.push(newRow);

    const scale: NumberedScale = {
        sign: [],
        number: [],
    };

    for(let i = left; i <= right; i += 1){

        getSigns(scale, i);
    }

    matrixArray.push(scale.number);
    matrixArray.push(scale.sign);

    matrixArray = matrixArray.reverse();
    return matrixArray;


};

const getScaleVertical = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const top  = matrix.getBound().topRight().y;
    const bottom = matrix.getBound().bottomLeft().y;

    const scale: NumberedScale = {
        sign: [' ', ' ', ' '],
        number: [' ', ' ', ' '],
    };

    for(let i = top; i >= bottom; i -= 1){

        getSigns(scale, i);

    }

    for(let i = 0; i < matrixArray.length; i += 1){

        matrixArray[i] = matrixArray[i].reverse();

        matrixArray[i].push(' ')
        matrixArray[i].push(' ')

        matrixArray[i].push(scale.number[i])
        matrixArray[i].push(scale.sign[i])

        matrixArray[i] = matrixArray[i].reverse();

    }

    return matrixArray;

};

const addScale = (matrix: Matrix, matrixString: string[][]): string[][] => {

    matrixString = getScaleHorizontal(matrix, matrixString);
    matrixString = getScaleVertical(matrix, matrixString);
    return matrixString;
};

export const matrixToString = (matrix: Matrix, scale = true): string => {

    if(0 === matrix.getAll().length){
        return 'Matrix is empty';
    }

    const width = matrix.getWidth();
    const height = matrix.getHeight();

    const topLeft = new Position(
        matrix.getBound().bottomLeft().x,
        matrix.getBound().topRight().y
    );

    let matrixOfStrings: string[][] = [];

    let pointer = topLeft.clone();

    for(let y = 0; y < height; y += 1){

        const line = [];

        for(let x = 0; x < width; x += 1){

            if(true === matrix.has(pointer)){
                line.push('O');
            }
            else{
                line.push('.');
            }

            pointer = pointer.move(new Position(1,0));

        }

        matrixOfStrings.push(line);

        pointer = new Position(
            topLeft.x,
            pointer.y - 1
        );

    }

    if(true === scale){
        matrixOfStrings = addScale(matrix, matrixOfStrings);
    }

    return matrixOfStrings.map((row)=>{return row.join('')}).join('\n');;

}
