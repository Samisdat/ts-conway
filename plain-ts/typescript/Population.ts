import {Position} from '@Conway/Geometry/Position';
import {Matrix} from '@Conway/Geometry/Matrix';

export class Population {

    private readonly generationDuration: number;

    private matrix: Matrix;

    constructor(
        matrix: Matrix,
        generationDuration: number
    ) {

        this.matrix = matrix;

        this.generationDuration = generationDuration;

    }

    public startAging(): void {

        setInterval(
            () => {
                this.elapse();
            },
            this.generationDuration
        );

    }

    private isLiving(position: Position): boolean {

        return this.matrix.has(position);

    }

    private countLivingNeighbours(position: Position): number {

        let livingNeighbours = 0;

        const neighbours: Position[] = position.getNeighbours();

        for (const neighbour of neighbours) {

            const isLiving = this.isLiving(neighbour);

            if (true === isLiving) {
                livingNeighbours += 1;
            }

        }

        return livingNeighbours;

    }

    private applyRules(): void {

        const nextGeneration = new Matrix();

        /**
         * there are four rules
         * 1) Any live cell with fewer than two live neighbours dies, as if caused by under-population.
         * 2) Any live cell with two or three live neighbours lives on to the next generation.
         * 3) Any live cell with more than three live neighbours dies, as if by over-population.
         * 4) Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
         *
         *  All rules dealing with living cells can be handled by kill all cells that don't have two or three neighbours
         *  That's the first loop
         *
         *  To check if any death cells will be birth, we count the living neighbour of all cells neighbours
         *  That's the second loop
         */

        for (const position of this.matrix.getAll()) {

            const livingNeighbours = this.countLivingNeighbours(position);

            if (2 === livingNeighbours || 3 === livingNeighbours) {

                nextGeneration.add(position);

            }

        }

        for (const position of this.matrix.getAll()) {

            const neighbours = position.getNeighbours();

            for (const neighbour of neighbours) {

                if (true === this.matrix.has(neighbour)) {
                    continue;
                }

                const livingNeighbours = this.countLivingNeighbours(neighbour);

                if (3 === livingNeighbours) {

                    nextGeneration.add(neighbour);

                }

            }

        }

        this.matrix = nextGeneration;

    }

    public elapse(): void {

        this.applyRules();

    }

    public getAllCells(): Position[] {

        return this.matrix.getAll();

    }

    get(): Position[] {

        return this.matrix.getAll();

    }

    public getMatrix(): Matrix {
        return this.matrix;
    }

}
