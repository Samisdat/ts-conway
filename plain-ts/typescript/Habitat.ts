import {Position} from '@Conway/Geometry/Position';
import {Population} from '@Conway/Population';
import {Cell} from '@Conway/Habitat/Cell';
import {CellTypesFactory} from '@Conway/Frontend/CellType/CellTypesFactory';
import {Dimension} from '@Conway/Habitat/Dimension';

interface MatrixCol {
    [index: number]: number;
}

interface MatrixRow {
    [index: number]: MatrixCol;
}

export class Habitat {

    private gridDimension: Dimension;
    private sourcePosition: Position;

    private offset: Position;

    private cells: Cell[] = [];

    public readonly center: Position;

    private absoluteMap: MatrixRow = {};

    constructor(
        population: Population,
        gridDimension: Dimension,
        sourcePosition: Position,
        offset: Position
    ) {

        Habitat.ensureOffsetBetweenMinusOneAndOne(offset);

        this.gridDimension = gridDimension;

        this.sourcePosition = sourcePosition;

        this.offset = offset;

        let relativePointer = new Position(0, 0);
        let absolutePointer = this.sourcePosition as Position;

        const inMiddleOfGrid = new Position(
            Math.floor(this.gridDimension.getRows() / 2),
            Math.floor(this.gridDimension.getCols() / 2)
        );

        this.center = relativePointer.move(inMiddleOfGrid);

        relativePointer = relativePointer.move(inMiddleOfGrid.inverse());

        absolutePointer = absolutePointer.move(inMiddleOfGrid.inverse());

        this.createGridCells(relativePointer, absolutePointer);

        this.markLivingCells(population);

    }

    private createGridCells(relativePointer: Position, absolutePointer: Position){

        const moveRight = new Position(
            1,
            0
        );

        // @TODO Use Bounds

        for (let top = 0; top < this.gridDimension.getCols(); top += 1) {

            for (let right = 0; right < this.gridDimension.getRows(); right += 1) {

                const gridCell = new Cell(
                    relativePointer,
                    absolutePointer,
                    this.offset
                );

                const absolutePosition = gridCell.absolutePosition;

                if (undefined === this.absoluteMap[absolutePosition.x]) {
                    this.absoluteMap[absolutePosition.x] = {};
                }

                this.absoluteMap[absolutePosition.x][absolutePosition.y] = this.cells.length;

                this.cells.push(gridCell);

                relativePointer = relativePointer.move(
                    moveRight
                );

                absolutePointer = absolutePointer.move(moveRight);

            }

            relativePointer = relativePointer.move(
                new Position(-1 * this.gridDimension.getRows(), 1),
            );

            absolutePointer = absolutePointer.move(
                new Position(-1 * this.gridDimension.getRows(), 1),
            );

        }

    }

    private markLivingCells(population: Population){

        for (const positionWithLivingCells of population.get()) {

            if (undefined === this.absoluteMap[positionWithLivingCells.x]) {
                continue;
            }
            if (undefined === this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y]) {
                continue;
            }

            const cellIndex = this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y];

            this.cells[cellIndex].setType(CellTypesFactory.get().living());

        }

    }

    private static ensureOffsetBetweenMinusOneAndOne(offset: Position) {

        if (1 < Math.abs(offset.x )) {
            throw new Error('offset.x must be between -1 and 1');
        }

        if (1 < Math.abs(offset.y)) {
            throw new Error('offset.y must be between -1 and 1');
        }


    }

    public getSourcePosition(): Position {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset;
    }

    public getCols(): number {
        return this.gridDimension.getCols();
    }

    public getRows(): number {
        return this.gridDimension.getRows();
    }

    public getCells() {
        return this.cells;
    }

}
