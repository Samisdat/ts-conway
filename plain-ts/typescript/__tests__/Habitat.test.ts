import {Population} from '@Conway/Population';
import {Dimension} from '@Conway/Habitat/Dimension';
import {Position} from '@Conway/Geometry/Position';
import {Matrix} from '@Conway/Geometry/Matrix';
import {Habitat} from '@Conway/Habitat';

describe('Habitat', () => {

    let population: Population;

    beforeAll(() => {

        const matrix = new Matrix();

        population = new Population(
            matrix,
            1000
        );

    });

    test('can be created', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 5),
            new Position(0, 0),
            new Position(0, 0)
        );

        expect(habitat).toBeInstanceOf(Habitat);

    });

    test('throw execption for offset.x below -1 or higher then 1', () => {

        expect(() => {

            new Habitat(
                population,
                new Dimension(2, 3),
                new Position(0, 0),
                new Position(-1.1, 0)
            );

        }).toThrowErrorMatchingSnapshot();

        expect(() => {

            new Habitat(
                population,
                new Dimension(2, 3),
                new Position(0, 0),
                new Position(1.1, 0)
            );

        }).toThrowErrorMatchingSnapshot();

    });

    test('throw execption for offset.y below -1 or higher then 1', () => {

        expect(() => {

            new Habitat(
                population,
                new Dimension(2, 3),
                new Position(0, 0),
                new Position(0, -1.1)
            );

        }).toThrowErrorMatchingSnapshot();

        expect(() => {

            new Habitat(
                population,
                new Dimension(2, 3),
                new Position(0, 0),
                new Position(0, 1.1)
            );

        }).toThrowErrorMatchingSnapshot();

    });


    test('get rows/cols odd number of rows/cols', () => {

        const habitat = new Habitat(
                population,
                new Dimension(3, 5),
                new Position(0, 0),
                new Position(0, 0)
            );

        expect(habitat.getRows()).toBe(3);
        expect(habitat.getCols()).toBe(5);

        expect(habitat.getSourcePosition()).toStrictEqual(
            new Position(0, 0)
        );

        expect(habitat.getOffset()).toStrictEqual(
            new Position(0, 0)
        );

        expect(habitat.getCells().length).toBe(3 * 5);

    });

    test('get rows/cols odd number of rows/cols with integer offset', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 5),
            new Position(-1, -2),
            new Position(0, 0)
        );

        expect(habitat.getRows()).toBe(3);
        expect(habitat.getCols()).toBe(5);

        expect(habitat.getSourcePosition()).toStrictEqual(
            new Position(-1, -2)
        );

        expect(habitat.getOffset()).toStrictEqual(
            new Position(0, 0)
        );

    });

    test('get rows/cols odd number of rows/cols with decimal offset', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 5),
            new Position(-1, -2),
            new Position(-0.5, -0.5)
        );

        expect(habitat.getRows()).toBe(3);
        expect(habitat.getCols()).toBe(5);

        expect(habitat.getSourcePosition()).toStrictEqual(
            new Position(-1, -2)
        );

        expect(habitat.getOffset()).toStrictEqual(
            new Position(-0.5, -0.5)
        );

    });

    test('get relative and absolute positions', () => {

        const habitat = new Habitat(
            population,
            new Dimension(1, 1),
            new Position(-1, -2),
            new Position(0, 0)
        );


        const cells = habitat.getCells();

        expect(cells.length).toBe(1);

        const cell = cells[0];

        expect(cell.relativePosition).toStrictEqual(
            new Position(0, 0)
        );

        expect(cell.absolutePosition).toStrictEqual(
            new Position(-1, -2)
        );

    });

    test('get cell with offset', () => {

        const habitat = new Habitat(
            population,
            new Dimension(1, 1),
            new Position(0, 0),
            new Position(-0.5, 0)
        );

        const cells = habitat.getCells();

        expect(cells.length).toBe(1);

        const cellOne = cells[0];

        expect(cellOne.relativePosition).toStrictEqual(
            new Position(0, 0)
        );

        expect(cellOne.x).toBe(-0.5);

        expect(cellOne.absolutePosition).toStrictEqual(
            new Position(0, 0)
        );

    });

    test('get cells positions without offset', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(0, 0),
            new Position(0, 0)
        );

        const cells = habitat.getCells();

        expect(cells.length).toBe(9);

        const relativePositions: Position[] = [];
        const absolutePositions: Position[] = [];

        for (const cell of cells) {
            relativePositions.push(cell.relativePosition);
            absolutePositions.push(cell.absolutePosition);
        }

        expect(relativePositions).toStrictEqual([
            new Position(-1, -1),
            new Position(0, -1),
            new Position(1, -1),
            new Position(-1, 0),
            new Position(0, 0),
            new Position(1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1)
        ]);

        // no offset, so relative and absolute positions should be the same
        expect(relativePositions).toStrictEqual(absolutePositions);

    });

    test('get cells positions with offset', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(0, 0),
            new Position(-0.5, -0.5)
        );

        const cells = habitat.getCells();

        expect(cells.length).toBe(9);

        const relativePositions: Position[] = [];
        const absolutePositions: Position[] = [];

        for (const cell of cells) {
            relativePositions.push(cell.relativePosition);
            absolutePositions.push(cell.absolutePosition);
        }

        expect(relativePositions).toStrictEqual([
            new Position(-1, -1),
            new Position(0, -1),
            new Position(1, -1),
            new Position(-1, 0),
            new Position(0, 0),
            new Position(1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1)
        ]);

        // no offset, so relative and absolute positions should be the same
        expect(relativePositions).toStrictEqual(absolutePositions);

    });

    test('get cells positions without offset 0/1', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(0, 1),
            new Position(0, 0)
        );

        const cells = habitat.getCells();

        expect(cells.length).toBe(9);

        const relativePositions: Position[] = [];
        const absolutePositions: Position[] = [];

        for (const cell of cells) {
            relativePositions.push(cell.relativePosition);
            absolutePositions.push(cell.absolutePosition);
        }

        expect(relativePositions).toStrictEqual([
            new Position(-1, -1),
            new Position(0, -1),
            new Position(1, -1),
            new Position(-1, 0),
            new Position(0, 0),
            new Position(1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1)
        ]);

        expect(absolutePositions).toStrictEqual([
            new Position(-1, 0),
            new Position(0, 0),
            new Position(1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1),
            new Position(-1, 2),
            new Position(0, 2),
            new Position(1, 2)
        ]);

    });

    test('get cells positions without offset 1/0', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(1, 0),
            new Position(0, 0)
        );

        const cells = habitat.getCells();

        expect(cells.length).toBe(9);

        const relativePositions: Position[] = [];
        const absolutePositions: Position[] = [];

        for (const cell of cells) {
            relativePositions.push(cell.relativePosition);
            absolutePositions.push(cell.absolutePosition);
        }

        expect(relativePositions).toStrictEqual([
            new Position(-1, -1),
            new Position(0, -1),
            new Position(1, -1),
            new Position(-1, 0),
            new Position(0, 0),
            new Position(1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1)
        ]);

        expect(absolutePositions).toStrictEqual([
            new Position(-1 + 1, -1),
            new Position(0 + 1, -1),
            new Position(1 + 1, -1),
            new Position(-1 + 1, 0),
            new Position(0 + 1, 0),
            new Position(1 + 1, 0),
            new Position(-1 + 1, 1),
            new Position(0 + 1, 1),
            new Position(1 + 1, 1)
        ]);

    });

    test('get cells coordinates without offset', () => {

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(0, 0),
            new Position(0, 0)
        );

        const coordinates: Position[] = [];

        for (const cell of habitat.getCells()) {
            coordinates.push(
                new Position(cell.x, cell.y)
            );
        }

        expect(coordinates).toStrictEqual([
            new Position(-1, -1),
            new Position(0, -1),
            new Position(1, -1),
            new Position(-1, 0),
            new Position(0, 0),
            new Position(1, 0),
            new Position(-1, 1),
            new Position(0, 1),
            new Position(1, 1)
        ]);

    });

    test('get cells coordinates without offset -0.5/0', () => {

        const xOffset = -0.5;

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(0, 0),
            new Position(xOffset, 0)
        );

        const coordinates: Position[] = [];

        for (const cell of habitat.getCells()) {
            coordinates.push(
                new Position(cell.x, cell.y)
            );
        }

        expect(coordinates).toStrictEqual([
            new Position(-1 + xOffset, -1),
            new Position(0 + xOffset, -1),
            new Position(1 + xOffset, -1),
            new Position(-1 + xOffset, 0),
            new Position(0 + xOffset, 0),
            new Position(1 + xOffset, 0),
            new Position(-1 + xOffset, 1),
            new Position(0 + xOffset, 1),
            new Position(1 + xOffset, 1)
        ]);

    });

    test('get cells coordinates without offset 0/-0.5', () => {

        const yOffset = -0.5;

        const habitat = new Habitat(
            population,
            new Dimension(3, 3),
            new Position(0, 0),
            new Position(0, yOffset)
        );

        const coordinates: Position[] = [];

        for (const cell of habitat.getCells()) {
            coordinates.push(
                new Position(cell.x, cell.y)
            );
        }

        expect(coordinates).toStrictEqual([
            new Position(-1, -1 + yOffset),
            new Position(0, -1 + yOffset),
            new Position(1, -1 + yOffset),
            new Position(-1, 0 + yOffset),
            new Position(0, 0 + yOffset),
            new Position(1, 0 + yOffset),
            new Position(-1, 1 + yOffset),
            new Position(0, 1 + yOffset),
            new Position(1, 1 + yOffset)
        ]);

    });

    test('relative center can be retrieved', () => {

        let habitat = new Habitat(
            population,
            new Dimension(1, 1),
            new Position(0, 0),
            new Position(0, 0)
        );

        expect(habitat.center).toStrictEqual(
            new Position(0, 0)
        );

        habitat = new Habitat(
            population,
            new Dimension(3, 1),
            new Position(0, 0),
            new Position(0, 0)
        );

        expect(habitat.center).toStrictEqual(
            new Position(1, 0)
        );

        habitat = new Habitat(
            population,
            new Dimension(1, 3),
            new Position(0, 0),
            new Position(0, 0)
        );

        expect(habitat.center).toStrictEqual(
            new Position(0, 1)
        );


        habitat = new Habitat(
            population,
            new Dimension(2, 2),
            new Position(0, 0),
            new Position(0, 0)
        );

        expect(habitat.center).toStrictEqual(
            new Position(1, 1)
        );

        habitat = new Habitat(
            population,
            new Dimension(4, 4),
            new Position(0, 0),
            new Position(0, 0)
        );

        expect(habitat.center).toStrictEqual(
            new Position(2, 2)
        );

    });

});
