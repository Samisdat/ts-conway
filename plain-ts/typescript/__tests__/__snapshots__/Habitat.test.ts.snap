// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habitat throw execption for offset.x below -1 or higher then 1 1`] = `"offset.x must be between -1 and 1"`;

exports[`Habitat throw execption for offset.x below -1 or higher then 1 2`] = `"offset.x must be between -1 and 1"`;

exports[`Habitat throw execption for offset.y below -1 or higher then 1 1`] = `"offset.y must be between -1 and 1"`;

exports[`Habitat throw execption for offset.y below -1 or higher then 1 2`] = `"offset.y must be between -1 and 1"`;
